CLI/Sources/KnitCodeGen/ConfigurationSet.swift

```diff
@@ -145,3 +145,56 @@ public extension ConfigurationSet {
                 """
 
 }
+
+// MARK: - Validation
+
+/// Validate that there are not duplicate registrations _within_ this ConfigurationSet
+/// which represents a single module.
+/// Note that this will not find duplicate registrations across modules.
+extension ConfigurationSet {
+
+    private struct Key: Hashable {
+        let service: String
+        let name: String?
+    }
+
+    public func validateNoDuplicateRegistrations() throws {
+        var registrationCount = 0
+        var registrationSet = Set<Key>()
+
+        try assemblies
+            // Get all registrations across all assemblies
+            .flatMap { assembly in
+                assembly.registrations
+            }
+            .forEach { registration in
+                registrationCount += 1
+                registrationSet.insert(Key(service: registration.service, name: registration.name))
+
+                guard registrationCount == registrationSet.count else {
+                    throw ConfigurationSetParsingError.detectedDuplicateRegistration(
+                        service: registration.service,
+                        name: registration.name
+                    )
+                }
```

Comment:
Nit: `registrationCount` could be removed by checking if the registrationSet contains the key already
```suggestion
	let key = Key(service: registration.service, name: registration.name)
	guard !registrationSet.contains(key) else {
		throw ConfigurationSetParsingError.detectedDuplicateRegistration(
			service: registration.service,
			name: registration.name
		)
	}
	registrationSet.insert(key)
```